### PR TITLE
Ej/maps list

### PR DESCRIPTION
**Pull request by lightlii**
https://github.com/digidem/mapeo-webmaps/issues/56

A couple of things to note:
- I had some issues with the rebase so I don't seem to have all the latest changes from develop in here, perhaps @ErikSin we can look into that together before  you review this properly
- Slight deviation from the requirements in #56 - Can't sort 'By latest modified' since we don't record that info, (as far as I can tell firebase doesn't record any additional metadata either but I might be wrong).
-  The issue mentions that @ErikSin was going to look into the preview for the thumbnail. I had a look through Mapbox's documentation but didn't see any way we can get a preview of a style from the mapbox API. It looks like the only way would be to use the [static map endpoint](https://docs.mapbox.com/api/maps/static-images/) but that might be overkill for our case, what do you think?

### CODE REVIEW

File: src/views/Home/index.tsx [hunk 1]
```diff
@@ -1,21 +1,37 @@
...
 
 export const HomeView = () => {
+  const [user] = useAuthState(auth)
+  const [sort, setSort] = useState<SortType>('createdAt')
+  const [sortDirection, setSortDirection] = useState<SortDirectionType>('asc')
+
+  const mapsRef = collection(db, `groups/${user?.uid || ''}/maps`)
+
+  const [mapsCollection = { docs: [] }, mapsLoading] = useCollection(query(mapsRef, orderBy(sort, 'desc')))
+  const maps = mapsCollection?.docs.map((map) => ({ ...map?.data(), id: map?.id }))
```
Comment (ErikSin) [inline src/views/Home/index.tsx]: memoize

File: src/views/Home/index.tsx [hunk 2]
```diff
@@ -34,16 +50,43 @@ export const HomeView = () => {
...
+        <input {...getInputProps()} />
+        {uploading || failedFiles?.length ? (
+          <Uploading progress={progress} />
+        ) : (
+          <div {...getRootProps({ className: 'dropzone' })}>
+            <DragDropOverlay active={isDragActive} />
+            {maps.length ? (
+              <Container alignItems="flex-end">
+                <SortToggle value={sort} onChange={handleSortChange} direction={sortDirection} />
+                <MapsList maps={maps} progress={{ id: 1, loading: uploading }} />
```
Comment (ErikSin) [inline src/views/Home/index.tsx]: Where is the sorting happening on the map list?

File: src/components/Header/index.tsx [hunk 3]
```diff
@@ -50,9 +50,13 @@ export const Header = ({ children, onClickAddMap }: HeaderProps) => {
     <HeaderWrapper>
       {children || (
         <Row padding="0 0 0 18px">
-          <Block>
-            <AddMapButton onClick={onClickAddMap} />
-          </Block>
+          {onClickAddMap ? (
```
Comment (ErikSin) [inline src/components/Header/index.tsx]: When is there no `onClickAddMap`?

File: src/views/Home/index.tsx [hunk 4]
```diff
@@ -34,16 +50,43 @@ export const HomeView = () => {
...
-      ) : (
-        <div {...getRootProps({ className: 'dropzone' })}>
-          <DragDropOverlay active={isDragActive} />
-          <NoMaps openDialog={open} getInputProps={getInputProps} isDragActive={isDragActive} />
-        </div>
-      )}
+      <Stack sx={{ overflowY: 'scroll' }}>
+        <input {...getInputProps()} />
+        {uploading || failedFiles?.length ? (
+          <Uploading progress={progress} />
```
Comment (ErikSin) [inline src/views/Home/index.tsx:60]: '`Current File` is not being updated: 
image

File: src/hooks/useCreateMap.ts [hunk 5]
```diff
@@ -4,68 +4,61 @@ import * as path from 'path'
 import * as md5 from 'js-md5'
```
Comment (ErikSin) [inline src/hooks/useCreateMap.ts]: I think the simplest solution here is to just copy and paste `useCreateMap.ts` from `develop`, assuming you didn't touch this file for this PR

File: src/components/MapItem/index.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import {
+  Box,
+  Card,
+  CardContent,
+  CircularProgress,
+  IconButton,
+  Link,
+  Stack,
+  Tooltip,
```
Comment (ErikSin) [inline src/components/MapItem/index.tsx]: Remove `IconButton` and `ToolTip`

File: src/components/MapItem/index.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  Typography,
+} from '@mui/material'
+import { collection, FieldValue, Timestamp } from 'firebase/firestore'
```
Comment (ErikSin) [inline src/components/MapItem/index.tsx]: Remove `FieldValue`

File: src/components/MapItem/index.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+export const MapItem = ({ id, title, description, createdAt }: MapItemProps) => {
+  const [user] = useAuthState(auth)
+  const observationsRef = user ? collection(db, `groups/${user.uid}/maps/${id}/observations`) : null
+  const [observations = [], observationsLoading] = useCollectionData(observationsRef)
+  const { formatMessage } = useIntl()
+
+  if (!user) return null
+  const dateFormatOptions = { dateStyle: 'medium' } as DateFormatOptionsType
+
+  const shareUrl = `${SHARE_URL_BASE}/${user?.uid}/maps/${id}`
```
Comment (ErikSin) [inline src/components/MapItem/index.tsx:24]: Is this working for you? I only have one set of data, and it isn't working for that

File: src/components/MapsList/index.tsx [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import { DocumentData } from 'firebase/firestore'
+import { Grow, Typography } from '@mui/material'
```
Comment (ErikSin) [inline src/components/MapsList/index.tsx]: remove line

**Review by ErikSin — CHANGES_REQUESTED** [review]
Looks pretty good. I'm a little confused as when the sort is actually happening. Am I correct that it is not happening right now?

File: src/views/Home/index.tsx (hunk 2 above)
Comment (lightlii) [reply src/views/Home/index.tsx]: the sorting is happening in the query, see the end of line 29 `query(mapsRef, orderBy(sort, 'desc')`

**Comment by lightlii** [comment]
All fixes done from comments here and I've moved most of the maps list logic (inc NoMaps view) into MapsList as we discussed yesterday. Ready to merge unless you see anything else :)

**Comment by lightlii** [comment]
Just remembered we said we'd move the sorting to happen locally to reduce queries, so this branch will get an extra commit...

File: src/components/MapsList/index.tsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  const [sortDirection, setSortDirection] = useState<SortDirectionType>('asc')
+
+  const mapsRef = collection(db, `groups/${user?.uid || ''}/maps`)
+
+  const [mapsCollection = { docs: [] }, mapsLoading] = useCollection(
+    query(mapsRef, orderBy(sort, sortDirection)),
+  )
+  const maps = useMemo(() => mapsCollection?.docs.map((map) => ({ ...map?.data(), id: map?.id })), [
+    mapsCollection,
+  ]) as DocumentData[]
```
Comment (ErikSin) [inline src/components/MapsList/index.tsx]: This is how you could sortDirectly in TS

````ts
 const [mapsCollection = { docs: [] }, mapsLoading] = useCollection(query(mapsRef)) // you can still add an inital sort here, but not necesarry

  const sortedMaps:DocumentData[] = useMemo(() => 
  {
    const unsorted:DocumentData[] = mapsCollection?.docs.map((map) => ({ ...map?.data(), id: map?.id }))

    if(sort === 'createdAt'){
      return unsorted.sort((a,b)=> sortDirection === 'asc' ? a.createdAt.toDate().getTime()-b.createdAt.toDate().getTime(): b.createdAt.toDate().getTime()-a.createdAt.toDate().getTime())
    }

    return unsorted.sort((a,b) => sortDirection === 'asc' ? a.title.localeCompare(b.title) : -1*a.title.localeCompare(b.title))
  }
  , [mapsCollection, sort, sortDirection])

File: src/components/MapsList/index.tsx [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  )
+}
+
+// type MapsListProps = {
+//   maps: DocumentData[]
+//   progress: {
+//     id?: number
+//     loading?: boolean
+//   }
+// }
```
Comment (ErikSin) [inline src/components/MapsList/index.tsx]: can remove

**Review by ErikSin — APPROVED** [review]
just a couple comments to resolve but can be merged

**Comment by lightlii** [comment]
Added the client-side sorting, can you have a quick look over it before I merge pls @ErikSin ? :)

**Review by ErikSin — APPROVED** [review]
Looks good!

**Pull request merged**